### ISSUE TITLE
Performance improvements

### ISSUE THREAD

**Issue by adlerhurst** [zitadel/zitadel#6440]
# Performance improvements

This issue combines multiple causes of performance issues which need to be analyzed.

## Quota enforcement

Described in [another issue](https://github.com/zitadel/zitadel/issues/6425)

The problem there is that the enforcement of quotas take too long, mostly caused by suboptimal queries of unaggregated data (OLAP).

The problem will be resolved by pre aggregating the data and using a read model (currently projection table) to query the quota object.

## Unused indexes

There are > 100 unused indexes, we should remove them.

Possible workflow
1. set indexes invisible
2. analyze performance
3. remove indexes

## endpoints

The endpoints analyzed represent the endpoints called most on zitadel.cloud

### /oauth/v2/keys endpoint

https://console.cloud.google.com/traces/list?project=zitadel-domains&start=1693129793982&end=1693133393982&tid=21ea39816ae3f73f12fa0f8a73253f04

### /oauth/v2/token endpoint

https://console.cloud.google.com/traces/list?project=zitadel-domains&start=1693129793982&end=1693133393982&tid=23c3c06e3bec930413f1c2b729ed4ad1

### /oauth/v2/introspect

https://console.cloud.google.com/traces/list?project=zitadel-domains&start=1693129793982&end=1693133393982&tid=1bbf31cb678b0e14088d92c547ab7614

### /.well-known/openid-configuration

https://console.cloud.google.com/traces/list?project=zitadel-domains&start=1693129793982&end=1693133393982&tid=16ccb0b85af9331e105372855890f6a1

### /login

https://console.cloud.google.com/traces/list?project=zitadel-domains&tid=13135996944ef10312a5a21ee3190180

### /password

https://console.cloud.google.com/traces/list?project=zitadel-domains&tid=10cb1a24afd4a15510626166f91349fa 

## Spans

Currently there are sometimes huge gaps between spans (e.g. https://console.cloud.google.com/traces/list?project=zitadel-domains&start=1693132977326&end=1693158019826&tid=00fead1b6c7d0d7612caa56582fee9e9&pageState=(%22traceIntervalPicker%22:(%22groupValue%22:%22PT1H%22,%22customValue%22:null)), https://console.cloud.google.com/traces/list?project=zitadel-domains&start=1693132977326&end=1693158019826&tid=1aee738b31a93cd212e2810f8b214289&pageState=(%22traceIntervalPicker%22:(%22groupValue%22:%22PT1H%22,%22customValue%22:null))) we should add more fine graned spans to understand what is happening during these times.

| name | description | report | expected p99 latency (ms) |
|---|---|---|---|
| setInstance | occures in each call at the beginning, it is used to determine which instance is used for this call.  | [link](https://console.cloud.google.com/traces/tasks/8019809013306535238?project=zitadel-domains)  | 5ms |
| authorize middleware | occures in the calls which need to check authn/authz of users. Depending on the tokens there are differnet sub spans. | [link](https://console.cloud.google.com/traces/tasks/3715179550456490919?project=zitadel-domains) | 20ms |
| nextSteps | occures during the login, it will be replaced as soon as the new login is used (TODO: check if correct). It calculates which page should be shown next to the end user. | [link](https://console.cloud.google.com/traces/tasks/15166846988878715197?project=zitadel-domains) |  |
| auth request by id | This call queries an auth request by id and tons of other data | [link](https://console.cloud.google.com/traces/tasks/14134271305462379618?project=zitadel-domains) | 5ms |

### authorize middleware

Possible resolutions there are caching the instance in memory, analyze why the query takes so long (InstanceByHost > 10ms < 40ms) 

#### Querying the same data multiple times

The middleware queries the same database tables multiple times requesting the same data in later steps.

Possible solutions:

- in memory cache (quick fix, caching should be the last resort)
- store the data in context.Context
- trigger each object only once per api call
- analyze long running spans
  - auth request: there are many policies read sequentially
  - query user
    - user by login name requires rework of the login names projection (maybe query, maybe table design)
    - user by id is too slow
  - user grants query is too slow
  - user metadata query is too slow
  - search project roles is too slow (e.g. op.GetClientByClientID)
- do we need the latest sequences in all those queries
- why can we see the checkAccess span on open endpoints (e.g. /keys)

**Comment by fforootd** [comment]
@adlerhurst should we not move this to [zitadel/devops](https://github.com/zitadel/devops)?

**Comment by adlerhurst** [comment]
> @adlerhurst should we not move this to [zitadel/devops](https://github.com/zitadel/devops)?
> 
> 
> 
> To me it looks like a mixed bag of challenges in the product and some of it in our cloud service :grin: 
> 
> 

Works for me
